scheduler/src/cook/mesos/unscheduled.clj

```diff
@@ -0,0 +1,157 @@
+  e.g. {:resources {:cpus 2 :mem 3} :constraints {\"unique_host_constraints\" 4}},
+  Returns a data structure suitable for presentation, e.g.
+  [{:reason \"Not enough CPU available\" :host_count 2}
+   {:reason \"Not enough Memory available\" :host_count 3}
+   {:reason \"Job already ran on this host.\" :host_count 4}]"
+  [raw-summary]
+  (reduce into [] [(map (fn [[k v]] {:reason (str "Not enough " k " available.")
+                                     :host_count v})
+                        (:resources raw-summary))
+                   (map (fn [[k v]] {:reason (or (constraint-name->message k) k)
+                                     :host_count v})
+                        (:constraints raw-summary))]))
+
+(defn check-fenzo-placement
+  "Places the job Under Investigation (next time Fenzo fails to place the job,
+  the details of that will be recorded).
+  If there are already any Fenzo placement failures associated with the job, returns
+  [(String explaining placement failure), {data structure containing host counts for
+  each type of failure}].
+  Otherwise, if job wasn't already under investigation,
+  returns [String indicating job is now under investigation, {}] "
+  [conn job]
+  (when-not (:job/under-investigation job)
+    @(d/transact conn [{:db/id (:db/id job)
+                        :job/under-investigation true}]))
+  (if (:job/last-fenzo-placement-failure job)
+    ["The job couldn't be placed on any available hosts."
+     {:reasons (-> job :job/last-fenzo-placement-failure edn/read-string
+                   fenzo-failures-for-user)}]
+    ["The job is now under investigation. Check back in a minute for more details!" {}]))
+
+(defn check-queue-position
+  "IFF the job is not first in the user's queue, returns
+  [\"You have x other jobs ahead in the queue\", {:jobs [other job uuids]]}]"
+  [conn job]
+  (let [db (d/db conn)
+        user (:job/user job)
+        job-uuid (:job/uuid job)
+        running-tasks (map
+                       (fn [j] (->> j
+                                    :job/instance
+                                    (filter util/instance-running?)
+                                    last))
+                       (util/jobs-by-user-and-state db user :job.state/running))
+        pending-tasks (map util/create-task-ent
+                           (util/jobs-by-user-and-state db user :job.state/waiting))
+        all-tasks (into running-tasks pending-tasks)
+        sorted-tasks (vec (sort (util/same-user-task-comparator) all-tasks))
+        queue-pos (first
+                   (keep-indexed
+                    (fn [i instance]
+                      (when (= (-> instance :job/_instance :job/uuid) job-uuid) i))
+                    sorted-tasks))
+        tasks-ahead (subvec sorted-tasks 0 queue-pos)]
+    (when (seq tasks-ahead)
+      [(str "You have " queue-pos " other jobs ahead in the queue.")
+       {:jobs (mapv #(-> % :job/_instance :job/uuid str) tasks-ahead)}])))
```

Comment:
Sure, if the user has more than 10 jobs ahead in his own queue, this list would be hard to absorb anyway. 👍